TestCacheManager.testCacheNames

```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:645)
    at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:638)
    at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:630)
    at com.hazelcast.spring.cache.TestCacheManager.testCacheNames(TestCacheManager.java:121)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK8/com.hazelcast$hazelcast-spring/558/testReport/junit/com.hazelcast.spring.cache/TestCacheManager/testCacheNames/

```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:764)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:757)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:749)
	at com.hazelcast.spring.cache.TestCacheManager.testCacheNames(TestCacheManager.java:126)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-ZuluJDK6/com.hazelcast$hazelcast-spring/160/testReport/junit/com.hazelcast.spring.cache/TestCacheManager/testCacheNames/

```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds , count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:764)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:757)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:749)
	at com.hazelcast.spring.cache.TestCacheManager.testCacheNames(TestCacheManager.java:126)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast-spring/225/testReport/junit/com.hazelcast.spring.cache/TestCacheManager/testCacheNames/

Another failure in a PR build run:
```
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds, count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:899)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:892)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:884)
	at com.hazelcast.spring.cache.TestCacheManager.testCacheNames(TestCacheManager.java:121)
```
https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/new-lab-fast-pr/8393/testReport/junit/com.hazelcast.spring.cache/TestCacheManager/testCacheNames/

--- a/hazelcast-spring/src/test/java/com/hazelcast/spring/CustomSpringJUnit4ClassRunner.java
+++ b/hazelcast-spring/src/test/java/com/hazelcast/spring/CustomSpringJUnit4ClassRunner.java
@@ -16,10 +16,15 @@
 
 package com.hazelcast.spring;
 
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.TestLoggingUtils;
 import org.junit.runners.model.InitializationError;
+import org.junit.runners.model.Statement;
 import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+import java.util.Set;
 
 public class CustomSpringJUnit4ClassRunner extends SpringJUnit4ClassRunner {
 
@@ -41,5 +46,23 @@ public class CustomSpringJUnit4ClassRunner extends SpringJUnit4ClassRunner {
      */
     public CustomSpringJUnit4ClassRunner(Class<?> clazz) throws InitializationError {
         super(clazz);
+    }
+
+    @Override
+    protected Statement withAfterClasses(Statement statement) {
+        final Statement originalStatement = super.withAfterClasses(statement);
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                originalStatement.evaluate();
+
+                Set<HazelcastInstance> instances = Hazelcast.getAllHazelcastInstances();
+                if (!instances.isEmpty()) {
+                    String message = "Instances haven't been shut down: " + instances;
+                    Hazelcast.shutdownAll();
+                    throw new IllegalStateException(message);
+                }
+            }
+        };
     }
 }

--- a/hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java
+++ b/hazelcast-spring/src/test/java/com/hazelcast/spring/cache/JCacheCacheManagerTest.java
@@ -20,6 +20,7 @@ import com.hazelcast.core.Hazelcast;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.spring.CustomSpringJUnit4ClassRunner;
 import com.hazelcast.test.annotation.QuickTest;
+import org.junit.AfterClass;
 import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
@@ -55,6 +56,7 @@ public class JCacheCacheManagerTest {
     private CacheManager cacheManager2;
 
     @BeforeClass
+    @AfterClass
     public static void start() {
         Hazelcast.shutdownAll();
     }

